### PR TITLE
MP Config fixes

### LINKED ISSUE #1698: MP Config - fix mutability for config sources
Helidon 2.0
- MP Config sources must be mutable by design (specification explicitly states that)
- any caching is done on source level, not on config level

This is currently in conflict with our Helidon SE model.
The following areas are problematic:

- we are casting to MP/SE, as we use the same implementation class - this may no longer be feasible
- we are using meta-config, which is designed for SE, will not work with mutable config source for MP
- we allow setup of MP server with a custom SE config, this should still work
- we are using SE config in our MP implementations, this should still work

### PR DESCRIPTION
**Pull request by tomas-langer**
MP Configuration is now compliant with the specification (mutability of config).
Please see #1698 for details.

Resolves #1698

### CODE REVIEW

**Comment by tomas-langer** [comment]
There is a problem with health and converter. Working on it.

**Comment by tomas-langer** [comment]
Also an issue with native image.

**Comment by tomas-langer** [comment]
Native image now works (tested all three MP examples).
Added built-in converters as we have in Helidon SE, to make sure our components work seamlessly.

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigBuilder.java [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        addBuiltIn(converters, LocalDate.class, ConfigMappers::toLocalDate);
+        addBuiltIn(converters, LocalDateTime.class, ConfigMappers::toLocalDateTime);
+        addBuiltIn(converters, LocalTime.class, ConfigMappers::toLocalTime);
+        addBuiltIn(converters, ZonedDateTime.class, ConfigMappers::toZonedDateTime);
+        addBuiltIn(converters, ZoneId.class, ConfigMappers::toZoneId);
+        addBuiltIn(converters, ZoneOffset.class, ConfigMappers::toZoneOffset);
+        addBuiltIn(converters, Instant.class, ConfigMappers::toInstant);
+        addBuiltIn(converters, OffsetTime.class, ConfigMappers::toOffsetTime);
+        addBuiltIn(converters, OffsetDateTime.class, ConfigMappers::toOffsetDateTime);
+        addBuiltIn(converters, YearMonth.class, YearMonth::parse); ;
```
Comment (aseovic) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpConfigBuilder.java:174]: Extra semicolon at the end is causing checkstyle to fail.

Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigBuilder.java]: fixed

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+        }
+
+        return Optional.empty();
+    }
+
+    private Optional<String> applyFilters(String propertyName, String stringValue) {
+        String result = stringValue;
+
+        for (MpConfigFilter filter : filters) {
+            result = filter.apply(propertyName, stringValue);
```
Comment (danielkec) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:232]: `filter.apply(propertyName, result)` or is it last filter wins?

File: config/config-mp/src/main/java/io/helidon/config/mp/MpEnvironmentVariablesSource.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Override
+    public Map<String, String> getProperties() {
+        return env;
+    }
+
+    @Override
+    public String getValue(String propertyName) {
+        // environment variable config source is immutable - we can safely cache all requested keys, so we
+        // not not execute the regular expression on every get
```
Comment (danielkec) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpEnvironmentVariablesSource.java:47]: do not

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java (hunk 2 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:232]: correct!

File: config/config-mp/src/main/java/io/helidon/config/mp/MpSystemPropertiesSource.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public Map<String, String> getProperties() {
+        Set<String> strings = props.stringPropertyNames();
+
+        Map<String, String> result = new HashMap<>();
+        strings.forEach(it -> result.put(it, props.getProperty(it)));
+        return result;
+    }
```
Comment (danielkec) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpSystemPropertiesSource.java:43]: `return (Map) props`

File: config/config/src/main/java/io/helidon/config/ConfigSourcesRuntime.java [hunk 5]
```diff
@@ -37,11 +37,11 @@
 final class ConfigSourcesRuntime {
     private final List<RuntimeWithData> loadedData = new LinkedList<>();
 
-    private List<ConfigSourceRuntimeBase> allSources;
-    private MergingStrategy mergingStrategy;
-    private Consumer<Optional<ObjectNode>> changeListener;
+    private final List<ConfigSourceRuntimeImpl> allSources;
+    private final MergingStrategy mergingStrategy;
+    private volatile Consumer<Optional<ObjectNode>> changeListener;
```
Comment (danielkec) [inline config/config/src/main/java/io/helidon/config/ConfigSourcesRuntime.java:42]: Is volatile necessary here? Value is assigned only in synchronized block

File: config/encryption/src/main/java/io/helidon/config/encryption/MpEncryptionFilter.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+/**
+ * Provides possibility to decrypt passwords from configuration sources.
+ * Configuration can be used to enforce encryption (e.g. we will fail on clear-text value).
+ * <p>
+ * Password in properties must be stored as follows:
+ * <ul>
+ * <li>${AES=base64} - encrypted password using a master password (must be provided to Prime through configuration, system
+ * property or environment variable)</li>
+ * <li>${RSA=base64} - encrypted password using a public key (private key must be available to Prime instance,
+ * its location must be provided to prime through configuration, system property or environment variable)</li>
```
Comment (danielkec) [inline config/encryption/src/main/java/io/helidon/config/encryption/MpEncryptionFilter.java:41]: prime/Prime

File: config/config-mp/src/main/java/io/helidon/config/mp/MpSystemPropertiesSource.java (hunk 4 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpSystemPropertiesSource.java:43]: I do not like to treat `Properties` as a string map, as it is in fact a `Map<String, Object>`. Not sure why, but that is the implementation...

File: config/config/src/main/java/io/helidon/config/ConfigSourcesRuntime.java (hunk 5 above)
Comment (tomas-langer) [reply config/config/src/main/java/io/helidon/config/ConfigSourcesRuntime.java:42]: It is not - the `.onChange() on line 86 is not in synchronized block.

File: config/encryption/src/main/java/io/helidon/config/encryption/MpEncryptionFilter.java (hunk 6 above)
Comment (tomas-langer) [reply config/encryption/src/main/java/io/helidon/config/encryption/MpEncryptionFilter.java:41]: Copy/pasted, wrong original, fixing both.

File: microprofile/tests/arquillian/src/main/java/io/helidon/microprofile/arquillian/HelidonDeployableContainer.java [hunk 7]
```diff
@@ -108,12 +106,25 @@ public void setup(HelidonContainerConfiguration configuration) {
...
+        }
     }
 
     @Override
     public void stop() {
         // No-op
+        if (null != dummyServer) {
+            dummyServer.stop();
+        }
     }
```
Comment (danielkec) [inline microprofile/tests/arquillian/src/main/java/io/helidon/microprofile/arquillian/HelidonDeployableContainer.java:128]: Not a noop any more

File: config/config/src/main/java/io/helidon/config/ConfigSourcesRuntime.java (hunk 5 above)
Comment (danielkec) [reply config/config/src/main/java/io/helidon/config/ConfigSourcesRuntime.java:42]: missed that

File: microprofile/tests/arquillian/src/main/java/io/helidon/microprofile/arquillian/HelidonDeployableContainer.java (hunk 7 above)
Comment (tomas-langer) [reply microprofile/tests/arquillian/src/main/java/io/helidon/microprofile/arquillian/HelidonDeployableContainer.java:128]: Fixed (and a few other code fixes in this class).

**Comment by barchetta** [comment]
You probably don't want to commit `examples/todo-app/frontend/package-lock.json`

**Pull request merged**